// uri-js/dist/esnext/uri.js
/**
 * URI.js
 *
 * @fileoverview An RFC 3986 compliant, scheme extendable URI parsing/validating/resolving library for JavaScript.
 * @author <a href="mailto:gary.court@gmail.com">Gary Court</a>
 * @see http://github.com/garycourt/uri-js
 */
/**
 * Copyright 2011 Gary Court. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are
 * permitted provided that the following conditions are met:
 *
 *    1. Redistributions of source code must retain the above copyright notice, this list of
 *       conditions and the following disclaimer.
 *
 *    2. Redistributions in binary form must reproduce the above copyright notice, this list
 *       of conditions and the following disclaimer in the documentation and/or other materials
 *       provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY GARY COURT ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL GARY COURT OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * The views and conclusions contained in the software and documentation are those of the
 * authors and should not be interpreted as representing official policies, either expressed
 * or implied, of Gary Court.
 */
import URI_PROTOCOL from "./regexps-uri";
import IRI_PROTOCOL from "./regexps-iri";
import punycode from "punycode";
import { toUpperCase, typeOf, assign } from "./util";
export const SCHEMES = {};
export function pctEncChar(chr) {
    const c = chr.charCodeAt(0);
    let e;
    if (c < 16)
        e = "%0" + c.toString(16).toUpperCase();
    else if (c < 128)
        e = "%" + c.toString(16).toUpperCase();
    else if (c < 2048)
        e = "%" + ((c >> 6) | 192).toString(16).toUpperCase() + "%" + ((c & 63) | 128).toString(16).toUpperCase();
    else
        e = "%" + ((c >> 12) | 224).toString(16).toUpperCase() + "%" + (((c >> 6) & 63) | 128).toString(16).toUpperCase() + "%" + ((c & 63) | 128).toString(16).toUpperCase();
    return e;
}
export function pctDecChars(str) {
    let newStr = "";
    let i = 0;
    const il = str.length;
    while (i < il) {
        const c = parseInt(str.substr(i + 1, 2), 16);
        if (c < 128) {
            newStr += String.fromCharCode(c);
            i += 3;
        }
        else if (c >= 194 && c < 224) {
            if ((il - i) >= 6) {
                const c2 = parseInt(str.substr(i + 4, 2), 16);
                newStr += String.fromCharCode(((c & 31) << 6) | (c2 & 63));
            }
            else {
                newStr += str.substr(i, 6);
            }
            i += 6;
        }
        else if (c >= 224) {
            if ((il - i) >= 9) {
                const c2 = parseInt(str.substr(i + 4, 2), 16);
                const c3 = parseInt(str.substr(i + 7, 2), 16);
                newStr += String.fromCharCode(((c & 15) << 12) | ((c2 & 63) << 6) | (c3 & 63));
            }
            else {
                newStr += str.substr(i, 9);
            }
            i += 9;
        }
        else {
            newStr += str.substr(i, 3);
            i += 3;
        }
    }
    return newStr;
}
function _normalizeComponentEncoding(components, protocol) {
    function decodeUnreserved(str) {
        const decStr = pctDecChars(str);
        return (!decStr.match(protocol.UNRESERVED) ? str : decStr);
    }
    if (components.scheme)
        components.scheme = String(components.scheme).replace(protocol.PCT_ENCODED, decodeUnreserved).toLowerCase().replace(protocol.NOT_SCHEME, "");
    if (components.userinfo !== undefined)
        components.userinfo = String(components.userinfo).replace(protocol.PCT_ENCODED, decodeUnreserved).replace(protocol.NOT_USERINFO, pctEncChar).replace(protocol.PCT_ENCODED, toUpperCase);
    if (components.host !== undefined)
        components.host = String(components.host).replace(protocol.PCT_ENCODED, decodeUnreserved).toLowerCase().replace(protocol.NOT_HOST, pctEncChar).replace(protocol.PCT_ENCODED, toUpperCase);
    if (components.path !== undefined)
        components.path = String(components.path).replace(protocol.PCT_ENCODED, decodeUnreserved).replace((components.scheme ? protocol.NOT_PATH : protocol.NOT_PATH_NOSCHEME), pctEncChar).replace(protocol.PCT_ENCODED, toUpperCase);
    if (components.query !== undefined)
        components.query = String(components.query).replace(protocol.PCT_ENCODED, decodeUnreserved).replace(protocol.NOT_QUERY, pctEncChar).replace(protocol.PCT_ENCODED, toUpperCase);
    if (components.fragment !== undefined)
        components.fragment = String(components.fragment).replace(protocol.PCT_ENCODED, decodeUnreserved).replace(protocol.NOT_FRAGMENT, pctEncChar).replace(protocol.PCT_ENCODED, toUpperCase);
    return components;
}
;
function _stripLeadingZeros(str) {
    return str.replace(/^0*(.*)/, "$1") || "0";
}
function _normalizeIPv4(host, protocol) {
    const matches = host.match(protocol.IPV4ADDRESS) || [];
    const [, address] = matches;
    if (address) {
        return address.split(".").map(_stripLeadingZeros).join(".");
    }
    else {
        return host;
    }
}
function _normalizeIPv6(host, protocol) {
    const matches = host.match(protocol.IPV6ADDRESS) || [];
    const [, address, zone] = matches;
    if (address) {
        const [last, first] = address.toLowerCase().split('::').reverse();
        const firstFields = first ? first.split(":").map(_stripLeadingZeros) : [];
        const lastFields = last.split(":").map(_stripLeadingZeros);
        const isLastFieldIPv4Address = protocol.IPV4ADDRESS.test(lastFields[lastFields.length - 1]);
        const fieldCount = isLastFieldIPv4Address ? 7 : 8;
        const lastFieldsStart = lastFields.length - fieldCount;
        const fields = Array(fieldCount);
        for (let x = 0; x < fieldCount; ++x) {
            fields[x] = firstFields[x] || lastFields[lastFieldsStart + x] || '';
        }
        if (isLastFieldIPv4Address) {
            fields[fieldCount - 1] = _normalizeIPv4(fields[fieldCount - 1], protocol);
        }
        const allZeroFields = fields.reduce((acc, field, index) => {
            if (!field || field === "0") {
                const lastLongest = acc[acc.length - 1];
                if (lastLongest && lastLongest.index + lastLongest.length === index) {
                    lastLongest.length++;
                }
                else {
                    acc.push({ index, length: 1 });
                }
            }
            return acc;
        }, []);
        const longestZeroFields = allZeroFields.sort((a, b) => b.length - a.length)[0];
        let newHost;
        if (longestZeroFields && longestZeroFields.length > 1) {
            const newFirst = fields.slice(0, longestZeroFields.index);
            const newLast = fields.slice(longestZeroFields.index + longestZeroFields.length);
            newHost = newFirst.join(":") + "::" + newLast.join(":");
        }
        else {
            newHost = fields.join(":");
        }
        if (zone) {
            newHost += "%" + zone;
        }
        return newHost;
    }
    else {
        return host;
    }
}
const URI_PARSE = /^(?:([^:\/?#]+):)?(?:\/\/((?:([^\/?#@]*)@)?(\[[^\/?#\]]+\]|[^\/?#:]*)(?:\:(\d*))?))?([^?#]*)(?:\?([^#]*))?(?:#((?:.|\n|\r)*))?/i;
const NO_MATCH_IS_UNDEFINED = ("").match(/(){0}/)[1] === undefined;
export function parse(uriString, options = {}) {
    const components = {};
    const protocol = (options.iri !== false ? IRI_PROTOCOL : URI_PROTOCOL);
    if (options.reference === "suffix")
        uriString = (options.scheme ? options.scheme + ":" : "") + "//" + uriString;
    const matches = uriString.match(URI_PARSE);
           ìûƒ(ÿÿÿÿ        Cƒ(    :ƒ(ÿÿÿÿ    "“   üƒ(   üûƒ(               ÿÿÿÿ4Bƒ(    ?Bƒ(   JBƒ("“   Tüƒ(                       ÿÿÿÿ÷Cƒ(ÿÿÿÿíCƒ("“   üƒ(                           è…(    «Ê€(             Äüƒ(ÿÿÿÿ        „Cƒ(    :ƒ(ÿÿÿÿ    "“   èüƒ(   Ôüƒ(                   è…(    sÚ€(            ,ıƒ("“   tıƒ(   <ıƒ(               ÿÿÿÿbDƒ(           jDƒ(   :ƒ(        ÿÿÿÿ?Dƒ("“   œıƒ(                           è…(    üÒ€(             Èıƒ(ÿÿÿÿ        Dƒ(    :ƒ(ÿÿÿÿ    "“   ìıƒ(   Øıƒ(                   è…(    ÈÏ€(             0şƒ(ÿÿÿÿ        ÊCƒ(    :ƒ(ÿÿÿÿ    "“   Tşƒ(   @şƒ(               ÿÿÿÿ®Bƒ("“   ˜şƒ(                       ÿÿÿÿÓBƒ("“   Äşƒ(                       ÿÿÿÿ§Cƒ(ÿÿÿÿ:ƒ("“   ğşƒ(                       ÿÿÿÿDƒ(ÿÿÿÿ:ƒ("“   $ÿƒ(                       ÿÿÿÿÀDƒ("“   Xÿƒ(                           è…(    Œñ€(             „ÿƒ(ÿÿÿÿ        óDƒ(    :ƒ(ÿÿÿÿ    "“   ¨ÿƒ(   ”ÿƒ(               ÿÿÿÿ¥Fƒ("“   ìÿƒ(                       ÿÿÿÿ¢Eƒ(    ­Eƒ(   ¸Eƒ(   ÃEƒ("“    „(                       ÿÿÿÿöEƒ(    Fƒ(   Fƒ("“   \ „(                       ÿÿÿÿrFƒ("“   ˜ „(                       ÿÿÿÿ?Fƒ("“   Ä „(                       "“   „(                       ÿÿÿÿ\Eƒ(    dEƒ(    oEƒ(ÿÿÿÿ:ƒ(ÿÿÿÿ    ÿÿÿÿ&Eƒ(   .Eƒ(   :ƒ(   6Eƒ(ÿÿÿÿ:ƒ(ÿÿÿÿAEƒ(
   IEƒ(
   :ƒ(
   QEƒ(ÿÿÿÿ:ƒ(ÿÿÿÿØFƒ(ÿÿÿÿ:ƒ("“   Œ„(                           è…(    -#(             À„(ÿÿÿÿ        5Gƒ(    :ƒ(ÿÿÿÿ    "“   ä„(   Ğ„(                   è…(    @!(             („(ÿÿÿÿ        Gƒ(    :ƒ(ÿÿÿÿ    "“   L„(   8„(               ÿÿÿÿXGƒ("“   „(                       ÿÿÿÿĞGƒ(ÿÿÿÿ:ƒ("“   ¼„(                       ÿÿÿÿ~Gƒ(ÿÿÿÿˆGƒ("“   ğ„(                           è…(    "4(             $„(ÿÿÿÿ        ­Gƒ(    :ƒ(ÿÿÿÿ    "“   H„(   4„(               ÿÿÿÿHƒ("“   Œ„(                       ÿÿÿÿ;Hƒ("“   ¸„(                           ÀR(    ô„(   „( „(<„(    H…(    ÿÿÿÿ       ĞQ(   p…(    ÿÿÿÿ       °Q(    Œ…(    ÿÿÿÿ       R("“
   |„(                       ÿÿÿÿ¼Hƒ(    ÄHƒ(    :ƒ(    ÌHƒ(    :ƒ(    ÔHƒ(    ÜHƒ(    :ƒ(    äHƒ(ÿÿÿÿ:ƒ("“   ğ„(                       ÿÿÿÿ^Hƒ(    iHƒ(   wHƒ(   …Hƒ(   “Hƒ("“   <„(                       ÿÿÿÿIƒ(ÿÿÿÿ:ƒ(    Iƒ(    :ƒ(    $Iƒ(    ,Iƒ(    :ƒ(    4Iƒ(    <Iƒ(    :ƒ(    DIƒ(ÿÿÿÿ:ƒ("“
   À„(                       ÿÿÿÿtIƒ(    |Iƒ(    „Iƒ(    ŒIƒ(    ”Iƒ(    œIƒ(    ¤Iƒ(    ¬Iƒ(    ´Iƒ(ÿÿÿÿ:ƒ(ÿÿÿÿÿIƒ("“   „(                       @           Š‡(ÿÿÿÿ    ÿÿÿÿ                  <„("“   L„(   \„(               ÿÿÿÿ$Jƒ(    /Jƒ("“   ”„(                       ÿÿÿÿbJƒ(ÿÿÿÿlJƒ(ÿÿÿÿvJƒ(ÿÿÿÿ€Jƒ("“   È„(                           è…(    À³(    è…(    w³(    è…(    ³(    è…(    ²(    è…(    /²("“   ä„(   €„(                            L„(            <„(   
         ,„(            „(            „(ÿÿÿÿ        ¥Jƒ(    :ƒ(ÿÿÿÿ    ÿÿÿÿ       ­Jƒ(   :ƒ(ÿÿÿÿ    ÿÿÿÿ       µJƒ(   :ƒ(ÿÿÿÿ    ÿÿÿÿ       ½Jƒ(   :ƒ(ÿÿÿÿ    ÿÿÿÿ       ÅJƒ(   :ƒ(ÿÿÿÿ    ÿÿÿÿèJƒ("“   „„(                       ÿÿÿÿKƒ(    Kƒ("“   °„(                       ÿÿÿÿiKƒ(    qKƒ(   |Kƒ("“   ä„(                       ÿÿÿÿ"Lƒ(ÿÿÿÿ:ƒ("“    	„(                       ÿÿÿÿôKƒ(    üKƒ("“   T	„(                       ÿÿÿÿÆLƒ(ÿÿÿÿ:ƒ("“   ˆ	„(                       ÿÿÿÿ|Lƒ("“   ¼	„(                       ÿÿÿÿ¡Lƒ("“   è	„(                       ÿÿÿÿOLƒ(ÿÿÿÿ:ƒ("“   
„(                       "“   l
„(                       ÿÿÿÿ¢Kƒ(    ªKƒ(   µKƒ(   ÀKƒ(   ËKƒ(ÿÿÿÿFKƒ(ÿÿÿÿ:ƒ("“   ”
„(                       ÿÿÿÿéLƒ("“   È
„(                       ÿÿÿÿMƒ("“   ô
„(                       ÿÿÿÿQMƒ(    \Mƒ("“    „(                           è…(    UD‚(             T„(ÿÿÿÿ        Nƒ(    :ƒ(ÿÿÿÿ    "“   x„(   d„(                   è…(    H‚(    è…(    æF‚("“   („(    „(               brackets and escape zone separator for IPv6
        uriTokens.push(_normalizeIPv6(_normalizeIPv4(String(components.host), protocol), protocol).replace(protocol.IPV6ADDRESS, (_, $1, $2) => "[" + $1 + ($2 ? "%25" + $2 : "") + "]"));
    }
    if (typeof components.port === "number" || typeof components.port === "string") {
        uriTokens.push(":");
        uriTokens.push(String(components.port));
    }
    return uriTokens.length ? uriTokens.join("") : undefined;
}
;
const RDS1 = /^\.\.?\//;
const RDS2 = /^\/\.(\/|$)/;
const RDS3 = /^\/\.\.(\/|$)/;
const RDS4 = /^\.\.?$/;
const RDS5 = /^\/?(?:.|\n)*?(?=\/|$)/;
export function removeDotSegments(input) {
    const output = [];
    while (input.length) {
        if (input.match(RDS1)) {
            input = input.replace(RDS1, "");
        }
        else if (input.match(RDS2)) {
            input = input.replace(RDS2, "/");
        }
        else if (input.match(RDS3)) {
            input = input.replace(RDS3, "/");
            output.pop();
        }
        else if (input === "." || input === "..") {
            input = "";
        }
        else {
            const im = input.match(RDS5);
            if (im) {
                const s = im[0];
                input = input.slice(s.length);
                output.push(s);
            }
            else {
                throw new Error("Unexpected dot segment condition");
            }
        }
    }
    return output.join("");
}
;
export function serialize(components, options = {}) {
    const protocol = (options.iri ? IRI_PROTOCOL : URI_PROTOCOL);
    const uriTokens = [];
    //find scheme handler
    const schemeHandler = SCHEMES[(options.scheme || components.scheme || "").toLowerCase()];
    //perform scheme specific serialization
    if (schemeHandler && schemeHandler.serialize)
        schemeHandler.serialize(components, options);
    if (components.host) {
        //if host component is an IPv6 address
        if (protocol.IPV6ADDRESS.test(components.host)) {
            //TODO: normalize IPv6 address as per RFC 5952
        }
        //if host component is a domain name
        else if (options.domainHost || (schemeHandler && schemeHandler.domainHost)) {
            //convert IDN via punycode
            try {
                components.host = (!options.iri ? punycode.toASCII(components.host.replace(protocol.PCT_ENCODED, pctDecChars).toLowerCase()) : punycode.toUnicode(components.host));
            }
            catch (e) {
                components.error = components.error || "Host's domain name can not be converted to " + (!options.iri ? "ASCII" : "Unicode") + " via punycode: " + e;
            }
        }
    }
    //normalize encoding
    _normalizeComponentEncoding(components, protocol);
    if (options.reference !== "suffix" && components.scheme) {
        uriTokens.push(components.scheme);
        uriTokens.push(":");
    }
    const authority = _recomposeAuthority(components, options);
    if (authority !== undefined) {
        if (options.reference !== "suffix") {
            uriTokens.push("//");
        }
        uriTokens.push(authority);
        if (components.path && components.path.charAt(0) !== "/") {
            uriTokens.push("/");
        }
    }
    if (components.path !== undefined) {
        let s = components.path;
        if (!options.absolutePath && (!schemeHandler || !schemeHandler.absolutePath)) {
            s = removeDotSegments(s);
        }
        if (authority === undefined) {
            s = s.replace(/^\/\//, "/%2F"); //don't allow the path to start with "//"
        }
        uriTokens.push(s);
    }
    if (components.query !== undefined) {
        uriTokens.push("?");
        uriTokens.push(components.query);
    }
    if (components.fragment !== undefined) {
        uriTokens.push("#");
        uriTokens.push(components.fragment);
    }
    return uriTokens.join(""); //merge tokens into a string
}
;
export function resolveComponents(base, relative, options = {}, skipNormalization) {
    const target = {};
    if (!skipNormalization) {
        base = parse(serialize(base, o ¾B ¿ ÀB Á Â‘B Ã‘ Ä’B Å’ Æ“B Ç“ È”B É” ÊüB Ëü Ì•B Í• Î–B Ï– Ğ¢B Ñ¢ Ò—B Ó— Ô˜B Õ˜ Ö™B ×™ ØšB Ùš Ú›B Û› ÜÁB İÁ Şœí	£ á! â! ãŸÃ	¥ æ   ¡ ¢  ¡ £ ç¤ è¥ é¦ï
	§ ì
§  ' ¨  ) ©  

²# íõ! îª! ï« ğ¬ ñ­ ò¢B ó¢ ôÕB õÕ ö®ï
	©¦ù¯  Ë °  

²# úõ! û± ü² ı³  Ğ ™B ş™ ÿB € ´	«¦„(µ  & ¶  ' ·  & ¸  ) ¹  & º  ‹ »  

²# …õ! †¼   ¡" ‡£" ˆ½! ‰¾! Š¿! ‹À ŒŠ! ğ
" Á Â Ã ‘Ä ’Å “Æ ”Ç •È  Ñ ®B –® —ÀB ˜À ™üB šü ›ıB œı ÁB Á Ÿİ	B  İ	 ¡B ¢ £Éï
	­¦¦+Ê   Ë  & Ì  ' Í   Î   Ï   Ğ  

²# §õ! ¨Ñ! ©Ò ªÓ «Ô ¬Õ ­Ö ®× ¯Ø °Ù ±Ú ²Û ³Ü ´İ µŞ ¶ß ·à ¸á ¹â ºã »ä  j åB ¼å ½æB ¾æ ¿çB Àç Á°B Â° ÃŞB ÄŞ ÅèB Æè ÇB È Ééê	°Å¨ëìíÌ)î  

ï   ­" Í­# Îğ   ¬" Ï¬# Ğå Ñç Òñ  

É ÓÉ Ôò  	ó Õó Öô  

õ ×õ Øö  ÷ Ù÷ Úø  7 ù Ûù ÜĞ İÂ Ş® ßÁ à³ áŠ â‚ ãú ä å« æ­ çû èü éı êş ëÿ ì€ íé	²‚ğƒ   „ ñ… ò… ó† ô† õ‡ ö‡ ÷ˆ! øû! ù‰ úŠ û‹ üÄ ıÅ şŒ ÿŞ	´Ö¦…&  +   +   &   & ‘  ) ’  Ó “   ”  

²# †õ! ‡• ˆ– ‰— Š˜ ‹™ Œš › œ   ‘Ÿ •÷   ø   ù   ú      » ®B –® —üB ˜ü ™ıB šı ›™B œ™ ¼B ¼ ŸB   ¡¡	¶¦¤i¢  + £  + ¤  + ¥  + ¦  + §  + ¨  + ©  + ª  + «  + ¬  & ­  & ®  & ¯  & °  & ±  & ²  ° ³  & ´  & µ  

²# ¥õ! ¦¡" §¶! ¨·! ©¸  
¹ ªº «» ¬¼ ­½ ®¾ ¯¿ °À ±Á ²Â ³Ã ´Ä µÅ ¶Æ ·Ç ¸È ¹É ºÊ »Ë ¼Ì ½Í ¾Î ¿Ï ÀĞ ÁÑ ÂÒ ÃÓ ÄÔ ÅÕ ÆÖ Ç× ÈØ ÉÙ ÊÚ ËÛ ÌÜ Íİ ÎŞ Ïß  ¨ B Ğ ÑB Ò ÓB Ô ÕB Ö ×àB Øà ÙáB Úá Û‘B Ü‘ İ’B Ş’ ß“B à“ á”B â” ãüB äü å¢B æ¢ ç—B è— é˜B ê˜ ë™B ì™ íšB îš ïâB ğâ ñ›B ò› óÁB ôÁ õB ö ÷ãï
	¸Öıä  z å  

²# şõ! ÿæ €ç è ‚é ‡÷   ø   ù   ú   ê  , €B ˆ€ ‰B Š ‹ëï
	ºÖ¦‘ì  ¥ í  ª î  ª ï  ª ğ  

²# ’õ! “ñ ”ò š÷   ø   ù   ú   ó  z ôB ›ô œõB õ öB Ÿö  ÷B ¡÷ ¢B £ ¤ø	¼Ö¦ª&ù   ú  ¨ û   ü  ° ı   ş  

²# «õ! ¬ÿ ­€ ® ¯‚ °ƒ ±„ ²… ³† ´‡ µˆ ¶‰ ·Š ¸‹ º÷   ø   ù   ú   Œ  T B » ¼±	B ½±	 ¾ÈB ¿È À²	B Á²	 ÂB Ã ÄB Å ÆÂ	¾ É 	¿Ö¦Ït‘  + ’  + “  + ”  + •   –   —   ˜  & ™  & š  & ›  & œ  )      ° Ÿ  Ä    ° ¡  ¨ ¢  Æ £   ¤  È ¥  ° ¦  Å §  Õ ¨  

²# Ğõ! Ñ©! Ò¡" Óª! Ô« Õ¬ Ö­ ×® Ø¯ Ù° Ú± Û² Ü³ İ² Ş³ ß´ àµ á¶ â· ã¸ ä¹ åº æ» ç¼ è½ é¾ ê¿ ëÀ ìÁ íÂ îÃ ïÄ ğÅ ñÆ òÇ óÈ ôÉ õÊ öË ÷Ì øÍ ùÎ úÏ ûĞ üÑ ıÒ ÿ÷   ø   ù   ú   Ó  Ö ÔB €Ô ÕB ‚Õ ƒÖB „Ö …×B †× ‡ØB ˆØ ‰ÙB ŠÙ ‹ÚB ŒÚ ÛB Û ÜB Ü ‘ØB ’Ø “İB ”İ •ÙB –Ù —ŞB ˜Ş ™ŞB šŞ ›ßB œß àB à ŸâB  â ¡ãB ¢ã £ßB ¤ß ¥B ¦ §à	ÁÖ¦­Xá   â  Ã ã  & ä  & å   æ  ° ç  ° è  ¨ é  ° ê  ¿ ë  Æ
